File: src/audioWfc/wfc/constraints/concepts/GenericCadenceSoftConstraint.java
package audioWfc.wfc.constraints.concepts;

import audioWfc.wfc.HigherValues;
import audioWfc.wfc.Tile;
import audioWfc.wfc.grabbers.Grabber;
import audioWfc.musicTheory.Key;
import audioWfc.musicTheory.Note;
import audioWfc.musicTheory.NoteUtils;
import audioWfc.musicTheory.chords.Chord;
import audioWfc.musicTheory.chords.ChordQuality;
import audioWfc.wfc.hierarchy.prototypes.Chordesque;


public abstract class GenericCadenceSoftConstraint extends SoftConstraint<Chordesque> {
    private Grabber<Integer> firstOffsetGrabber;
    private Grabber<Integer> secondOffsetGrabber;
    private Grabber<ChordQuality> firstQualityGrabber;
    private Grabber<ChordQuality> secondQualityGrabber;
    protected Grabber<Key> keyGrabber;

    public GenericCadenceSoftConstraint(double factor, Grabber<Key> keyGrabber,
                                        Grabber<Integer> firstOffsetGrabber, Grabber<ChordQuality> firstQualityGrabber,
                                        Grabber<Integer> secondOffsetGrabber, Grabber<ChordQuality> secondQualityGrabber){
        super(factor);
        this.keyGrabber = keyGrabber;
        this.firstOffsetGrabber = firstOffsetGrabber;
        this.secondOffsetGrabber = secondOffsetGrabber;
        this.firstQualityGrabber = firstQualityGrabber;
        this.secondQualityGrabber = secondQualityGrabber;
    }

    @Override
    public double weight(Tile<Chordesque> tile, HigherValues higherValues) {
        Chord chord = tile.getValue().getValue();

        Note root = keyGrabber.grab(higherValues).getRoot();

        Note firstRoot = NoteUtils.relativeNote(root, firstOffsetGrabber.grab(higherValues));
        Chord firstChord = Chord.create(firstRoot, firstQualityGrabber.grab(higherValues));

        Note secondRoot = NoteUtils.relativeNote(root, secondOffsetGrabber.grab(higherValues));
        Chord secondChord = Chord.create(secondRoot, secondQualityGrabber.grab(higherValues));

        if(chord.equals(secondChord)){
            Tile<Chordesque> prev = tile.getPrev();
            if(prev.isCollapsed() && prev.getValue().getValue().equals(firstChord)){
                return factor;
            }
        } else if(chord.equals(firstChord)){
            Tile<Chordesque> next = tile.getNext();
            if(next.isCollapsed() && next.getValue().getValue().equals(secondChord)){
                return factor;
            }
        }
        return 0d;
    }
}

File: src/wfc/constraints/cadences/GenericCadenceSoftConstraint.ts
import { ChordQuality, Chord } from "../../../music_theory/Chord"
import { MusicalKey } from "../../../music_theory/MusicalKey"
import { relativeNote } from "../../../music_theory/Note"
import { Grabber } from "../../Grabber"
import { HigherValues } from "../../HigherValues"
import { Tile } from "../../Tile"
import { Chordesque } from "../../hierarchy/Chordesque"
import { SoftConstraint } from "../concepts/Constraint"

export abstract class GenericCadenceSoftConstraint extends SoftConstraint<Chordesque> {
	protected firstOffsetGrabber: Grabber<number>
	protected secondOffsetGrabber: Grabber<number>
	protected firstChordGrabber: Grabber<ChordQuality>
	protected secondChordGrabber: Grabber<ChordQuality>
	protected keyGrabber: Grabber<MusicalKey>

	constructor(
		bonus: number,
		firstOffsetGrabber: Grabber<number>,
		secondOffsetGrabber: Grabber<number>,
		firstChordGrabber: Grabber<ChordQuality>,
		secondChordGrabber: Grabber<ChordQuality>,
		keyGrabber: Grabber<MusicalKey>,
		private reachOverPrev: boolean,
		private reachOverNext: boolean
	) {
		super(bonus)
		this.firstOffsetGrabber = firstOffsetGrabber
		this.secondOffsetGrabber = secondOffsetGrabber
		this.firstChordGrabber = firstChordGrabber
		this.secondChordGrabber = secondChordGrabber
		this.keyGrabber = keyGrabber
	}

	weight(tile: Tile<Chordesque>, higherValues: HigherValues): number {
		const chord = tile.getValue().getChord()

		const rootOfKey = this.keyGrabber(higherValues).getRoot()

		const firstRoot = relativeNote(
			rootOfKey,
			this.firstOffsetGrabber(higherValues),
		)
		const firstChord = Chord.fromRootAndQuality(
			firstRoot,
			this.firstChordGrabber(higherValues),
		)

		const secondRoot = relativeNote(
			rootOfKey,
			this.secondOffsetGrabber(higherValues),
		)
		const secondChord = Chord.fromRootAndQuality(
			secondRoot,
			this.secondChordGrabber(higherValues),
		)

		if (chord.equals(secondChord)) {
			const prev = tile.getPrev(this.reachOverPrev)
			if (
				prev.isCollapsed() &&
				prev.getValue().getChord() == firstChord
			) {
				return this.bonus
			}
		} else if (chord.equals(firstChord)) {
			const next = tile.getNext(this.reachOverNext)
			if (
				next.isCollapsed() &&
				next.getValue().getChord() == secondChord
			) {
				return this.bonus
			}
		}
		return 0
	}
}
